File: asml/LC54SpiralMatrix.py
class Solution(object):
    def spiralOrder(self, matrix):
        if not matrix or not matrix[0]:
            return []
        row_up = 0
        row_down = len(matrix)-1
        col_up = 0
        col_down = len(matrix[0])-1
        answer = []
        while row_up <= row_down and col_up <= col_down:
            i = col_up
            while i <= col_down:
                answer.append(matrix[row_up][i])
                i += 1
            row_up += 1
            i = row_up
            while i <= row_down:
                answer.append(matrix[i][col_down])
                i += 1
            col_down -= 1
            if row_up <= row_down:
                i = col_down
                while i >= col_up:
                    answer.append(matrix[row_down][i])
                    i -= 1
                row_down -= 1
            if col_up <= col_down:
                i = row_down
                while i >= row_up:
                    answer.append(matrix[i][col_up])
                    i -= 1
                col_up += 1
        return answer

File: asml/LC54SpiralMatrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        if (matrix.size() == 0 || matrix[0].size() == 0) return {};
        int row_up = 0;
        int row_down = matrix.size()-1;
        int col_up = 0;
        int col_down = matrix[0].size()-1;
        int k = 0;
        vector<int> answer(matrix.size()*matrix[0].size());
        while ((row_up <= row_down) && (col_up <= col_down)) {
            int i = col_up;
            while (i <= col_down) {
                answer[k++] = matrix[row_up][i];
                i++;
            }
            row_up++;
            i = row_up;
            while (i <= row_down) {
                answer[k++] = matrix[i][col_down];
                i++;
            }
            col_down--;
            if (row_up <= row_down) {
                i = col_down;
                while (i >= col_up) {
                    answer[k++] = matrix[row_down][i];
                    i--;
                }
                row_down--;
            }
            if (col_up <= col_down) {
                i = row_down;
                while (i >= row_up) {
                    answer[k++] = matrix[i][col_up];
                    i--;
                }
                col_up++;
            }
        }
        return answer;
    }
};
